src/Build/Utilities/StringPool.cs

```diff
@@ -0,0 +1,294 @@
+    public string Intern(ReadOnlySpan<char> span)
+    {
+        if (_buckets is null)
+        {
+            const int initialSize = 3; // must be prime
+            _buckets = new int[initialSize];
+            _slots = new Slot[initialSize];
+        }
+
+        int hashCode = InternalGetHashCode(span);
+
+        int bucketIndex = hashCode % _buckets.Length;
+
+        // Search for an existing entry.
+        for (int probeIndex = _buckets[hashCode % _buckets.Length] - 1; probeIndex >= 0;)
+        {
+            ref Slot slot = ref _slots![probeIndex];
+
+            if (slot.HashCode == hashCode && InternalEquals(slot.Value, span))
+            {
+                // Found a match! Return it.
+                return slot.Value;
+            }
+
+            // Follow the chain to find the next item.
+            probeIndex = slot.Next;
+        }
+
+        // We will add a new entry.
+        // Resize our storage if needed.
+        if (_count == _slots!.Length)
+        {
+            IncreaseCapacity();
+            bucketIndex = hashCode % _buckets.Length;
+        }
+
+        // Add the new entry in the last slot.
+        int slotIndex = _count++;
+
+        // Materialize a string for the span.
+        string str = span.ToString();
+
+        // Store the string.
+        _slots[slotIndex].HashCode = hashCode;
+        _slots[slotIndex].Value = str;
+        _slots[slotIndex].Next = _buckets[bucketIndex] - 1;
+        _buckets[bucketIndex] = slotIndex + 1;
+
+        // Return the newly created string.
+        return str;
+
+        void IncreaseCapacity()
+        {
+            int newSize = HashHelpers.ExpandPrime(_count);
+
+            if (newSize <= _count)
+            {
+                throw new OverflowException("StringPool size overflowed.");
+            }
+
+            Slot[] newSlots = new Slot[newSize];
+
+            if (_slots != null)
+            {
+                Array.Copy(_slots, 0, newSlots, 0, _count);
+            }
+
+            int[] newBuckets = new int[newSize];
+
+            for (int i = 0; i < _count; i++)
+            {
+                int num = newSlots[i].HashCode % newSize;
+                newSlots[i].Next = newBuckets[num] - 1;
+                newBuckets[num] = i + 1;
+            }
+
+            _slots = newSlots;
+            _buckets = newBuckets;
+        }
+    }
+
+    /// <summary>
+    /// Determines whether the provided string has equivalent content to the characters
+    /// in the provided span, using ordinal comparison.
+    /// </summary>
+    internal static unsafe bool InternalEquals(string str, ReadOnlySpan<char> span)
+    {
+        if (str.Length != span.Length)
+        {
+            return false;
+        }
+
+        // Walk both the string and the span
+        fixed (char* pStr0 = str)
+        {
+            fixed (char* pSpan0 = span)
+            {
+                // Reinterpret the characters (int16) as int (int32) so that we can
```

Comment:
"int16" -> UTF-16?

Reply:
Okay, I understand what you mean here, just took me a couple of reads.

Reply:
How about this?

```suggestion
                // Reinterpret the characters (16 bit) as int (32 bit) so that we can
```